File: Leetcode100/leetcode100/src/main/java/com/jiang/leetcode100/find_all_anagrams_in_a_string/FindAllAnagramsinaString.java
package main.java.com.jiang.leetcode100.find_all_anagrams_in_a_string;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

class Solution {
    public List<Integer> findAnagrams(String s, String p) {
        int[] map = new int[26];
        for(int i = 0; i < p.length(); i++) map[p.charAt(i) - 'a']++;

        List<Integer> result = new ArrayList<>();
        int i = 0;
        while(i < s.length()) {
            while(i < s.length() && map[s.charAt(i) - 'a'] == 0) i++;
            int[] temp = map.clone();
            int count = 0;
            int j = i;
            while(j < s.length() && temp[s.charAt(j) - 'a'] != 0) {
                temp[s.charAt(j) - 'a']--;
                count++;
                j++;
            }
            if(count == p.length())  result.add(i);
            i++;
        }
        return result;
    }
}

public class FindAllAnagramsinaString {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        String s = scanner.nextLine();
        String p = scanner.nextLine();
        scanner.close();

        List<Integer> result = new Solution().findAnagrams(s, p);
        for(Integer i : result) {
            System.out.print(i + " ");
        }
    }
}

File: Leetcode100/leetcode100_python/FindAllAnagramsinaString.py
from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(s) < len(p):
            return []
        m1 = {}
        m2 = {}
        for ch in p:
            m1[ch] = m1.get(ch, 0) + 1
        
        result = []
        left, right = 0, 0
        count = 0
        while right < len(s):
            if s[right] in m1 and m2.get(s[right], 0) < m1[s[right]]:
                m2[s[right]] = m2.get(s[right], 0) + 1
                count += 1
                right += 1
            else:
                if s[right] not in m1:
                    left = right + 1
                    count = 0
                    m2.clear()
                    right += 1
                else:
                    m2[s[left]] -= 1
                    count -= 1
                    left += 1
            if count == len(p):
                result.append(left)
        return result


if __name__ == "__main__":
    s = input()
    p = input()
    solution = Solution()
    print(solution.findAnagrams(s, p))